File: RenetikObjc/Library/Classes/Core/Classes/Network/CSConcurrentResponse.m
//
//  Created by Rene Dohan on 1/12/13.
//


#import "CSResponse.h"
#import "CSConcurrentResponse.h"
#import "NSObject+CSExtension.h"
#import "NSArray+CSExtension.h"
#import "NSMutableArray+CSExtension.h"
#import "CSLang.h"


@implementation CSConcurrentResponse {
    NSMutableArray *_failedResponses;
    NSMutableArray *_responses;
}

- (instancetype)init {
    if (self = super.init) {
        _failedResponses = NSMutableArray.new;
        _responses = NSMutableArray.new;
        self.data = NSMutableArray.new;
    }
    return self;
}

- (instancetype)addAll:(NSArray *)responses {
    for (CSResponse *response in responses) [self add:response];
    return self;
}

- (CSResponse *)add:(CSResponse *)response {
    [self.data put:response.data];
    __weak CSResponse *wResponse = [_responses put:response];
    return [[response onFailed:^(CSResponse *failedResponse) {[self onResponseFailed:wResponse];}]
            onSuccess:^(id data) {[self onResponseSuccess:wResponse];}];
}

- (void)onResponseSuccess:(CSResponse *)response {
    if ([_responses remove:response].empty) [self onResponsesDone];
}

- (void)onResponseFailed:(CSResponse *)failedResponse {
    if ([_responses remove:[_failedResponses put:failedResponse]].empty) [self onResponsesDone];
}

- (void)cancel {
    for (CSResponse *response in _responses) [response cancel];
    super.cancel;
}

- (void)onAddDone {
    invoke(^{if (_responses.empty)[self onResponsesDone];});
}

- (void)onResponsesDone {
    if (_failedResponses.hasItems) [self failed:_failedResponses.first];
    else [self success:self.data];
}

@end

File: Renetik/Renetik/Classes/Core/Classes/Response/CSConcurrentResponse.swift
//
//  Renetik Software
//
//  Created by Rene Dohan on 3/14/19.
//

import RenetikObjc

struct Test {

}

public class CSConcurrentResponse: CSResponse<NSMutableArray> {

    var failedResponse: CSResponseProtocol?
    let responses = CSArray<CSResponseProtocol>() // Cannot use swift array because of equatable shit

    public override init() { super.init(); self.data = NSMutableArray() }

    public init<T: AnyObject>(_ response: CSResponse<T>) { super.init(); add(response) }

    @discardableResult
    public func add<T: AnyObject>(_ response: CSResponse<T>) -> CSResponse<T> {
        data!.add(response.data!)
        responses.add(response)
        response.onFailed { _ in self.onResponseFailed(response) }
                .onSuccess { _ in self.onResponseSuccess(response) }
        return response
    }

    public func onAddDone() { later { if self.responses.isEmpty { self.success(self.data) } } }

    func onResponseSuccess<T: AnyObject>(_ response: CSResponse<T>) {
        responses.remove(response)
        if responses.isEmpty { self.success(self.data) }
    }

    func onResponseFailed<T: AnyObject>(_ failedResponse: CSResponse<T>) {
        self.failedResponse = failedResponse
        responses.remove(failedResponse).forEach { $0.cancel() }
        failed(failedResponse)
    }

    override open func cancel() { responses.forEach { $0.cancel() }; super.cancel() }
}